helper/resource/state.go

```diff
@@ -132,8 +134,42 @@ func (conf *StateChangeConf) WaitForState() (interface{}, error) {
 	case <-doneCh:
 		return result, resulterr
 	case <-time.After(conf.Timeout):
-		return nil, fmt.Errorf(
-			"timeout while waiting for state to become '%s'",
-			conf.Target)
+		return nil, ErrTimeout{
+			ExpectedState: conf.Target,
+			LastError:     resulterr,
+		}
 	}
 }
+
+// ErrResourceNotFound is returned when NotFoundChecks is set and the number of attempts is exceeded
+var ErrResourceNotFound = errors.New("couldn't find resource")
+
+// ErrUnexpectedState is returned when Refresh returns a state that's neither in Target nor Pending
+type ErrUnexpectedState struct {
```

Comment:
I agree we should introduce errors like this, let me just think about this concept more and try to compare pros/cons and use cases across more resources. AWS SDK is using slightly different model where they have one error `type` and then they use `err.Code()` for different kinds of errors.

I'm not sure (yet) which model would work best for Terraform.